12-offer-encoding.md

```diff
@@ -0,0 +1,894 @@
+                 
+Assume L1A < L2A:
+
+       L1A   L2A                                 L3A=H("LnBranch",L3nonce||L3)
+         \   /                                    |
+          v v                                     v
+  L1A2A=H("LnBranch",L1A||L2A)                   L3A=H("LnBranch",L3nonce||L3)
+  
+Assume L1A2A > L3A:
+
+  L1A2A=H("LnBranch",L1A||L2A)          L3A
+                          \            /
+                           v          v
+                Root=H("LnBranch",L3A||L1A2A)
+
+Signature = SIG("lightninginvoicesignature", Root, nodekey)
+```
+
+# Offers
+
+Offers are a precursor to an invoice_request: readers will request an invoice
+(or multiple) based on the offer.  An offer can be much longer-lived than a
+particular invoice, so it has some different characteristics; in particular the amount can be in a non-lightning currency.  It's
+also designed for compactness to fit inside a QR code easily.
+
+Note that the non-signature TLV elements get mirrored into
+invoice_request and invoice messages, so they each have specific and
+distinct TLV ranges.
+
+The human-readable prefix for offers is `lno`.
+
+## TLV Fields for Offers
+
+1. `tlv_stream`: `offer`
+2. types:
+    1. type: 2 (`offer_chains`)
+    2. data:
+        * [`...*chain_hash`:`chains`]
+    1. type: 4 (`offer_metadata`)
+    2. data:
+        * [`...*byte`:`data`]
+    1. type: 6 (`offer_currency`)
+    2. data:
+        * [`...*utf8`:`iso4217`]
+    1. type: 8 (`offer_amount`)
+    2. data:
+        * [`tu64`:`amount`]
+    1. type: 10 (`offer_description`)
+    2. data:
+        * [`...*utf8`:`description`]
+    1. type: 12 (`offer_features`)
+    2. data:
+        * [`...*byte`:`features`]
+    1. type: 14 (`offer_absolute_expiry`)
+    2. data:
+        * [`tu64`:`seconds_from_epoch`]
+    1. type: 16 (`offer_paths`)
+    2. data:
+        * [`...*blinded_path`:`paths`]
+    1. type: 18 (`offer_issuer`)
+    2. data:
+        * [`...*utf8`:`issuer`]
+    1. type: 20 (`offer_quantity_max`)
+    2. data:
+        * [`tu64`:`max`]
+    1. type: 22 (`offer_issuer_id`)
+    2. data:
+        * [`point`:`id`]
+
+## Requirements For Offers
+
+A writer of an offer:
+  - MUST NOT set any TLV fields outside the inclusive ranges: 1 to 79 and 1000000000 to 1999999999.
+  - if the chain for the invoice is not solely bitcoin:
+    - MUST specify `offer_chains` the offer is valid for.
+  - otherwise:
+    - SHOULD omit `offer_chains`, implying that bitcoin is only chain.
+  - if a specific minimum `offer_amount` is required for successful payment:
+    - MUST set `offer_amount` to the amount expected (per item).
+    - if the currency for `offer_amount` is that of all entries in `chains`:
+      - MUST specify `offer_amount` in multiples of the minimum lightning-payable unit
+        (e.g. milli-satoshis for bitcoin).
+    - otherwise:
+      - MUST specify `offer_currency` `iso4217` as an ISO 4712 three-letter code.
```

Comment:
(sorry)
```suggestion
      - MUST specify `offer_currency` `iso4217` as an ISO 4217 three-letter code.
```

Reply:
Damn, not ISO 4720: Nomenclature for Essential Oils?